app/scripts/api.js

```diff
@@ -14,622 +15,622 @@ import {
+       * genomic range (if chrom-sizes are available)
+       *
+       * .. code-block:: javascript
+       *
+       *  // Global output
+       *  {
+       *    dataRange: [...]
+       *    genomicRange: [...]
+       *  }
+       *
+       *  // 1D data range
+       *  [[1218210862, 1528541001], null]
+       *
+       *  // 2D data range
+       *  [[1218210862, 1528541001], [1218210862, 1528541001]]
+       *
+       *  // 1D or BED-like array
+       *  [["chr1", 249200621, "chrM", 50000], null]
+       *
+       *  // 2D or BEDPE-like array
+       *  [["chr1", 249200621, "chr2", 50000], ["chr3", 197972430, "chr4", 50000]]
+       *
+       * ``viewConfig:`` Returns the current view config.
+       *
+       * ``mouseMoveZoom:`` Returns the raw data around the mouse cursors screen location
+       * and the related genomic location.
+       *
+       * .. code-block:: javascript
+       *
+       *  {
+       *    data, // Raw Float32Array
+       *    dim,  // Dimension of the lens (the lens is squared)
+       *    toRgb,  // Current float-to-rgb converter
+       *    center,  // BED array of the cursors genomic location
+       *    xRange,  // BEDPE array of the x genomic range
+       *    yRange,  // BEDPE array of the y genomic range
+       *    rel  // If true the above three genomic locations are relative
+       *  }
+       *
+       * @param {string} event One of the events described below
+       *
+       * @param {function} callback A callback to be called when the event occurs
+       *
+       * @param {string} viewId The view ID to listen to events
+       *
+       * @example
+       *
+       *  let locationListenerId;
+       * hgv.on(
+       *   'location',
+       *   location => console.log('Here we are:', location),
+       *   'viewId1',
+       *   listenerId => locationListenerId = listenerId
+       * );
+       *
+       * const rangeListenerId = hgv.on(
+       *   'rangeSelection',
+       *   range => console.log('Selected', range)
+       * );
+       *
+       * const viewConfigListenerId = hgv.on(
+       *   'viewConfig',
+       *   range => console.log('Selected', range)
+       * );
+       *
+       *  const mmz = event => console.log('Moved', event);
+       *  hgv.on('mouseMoveZoom', mmz);
+       */
+      on(event, callback, viewId, callbackId) {
+        switch (event) {
+          case 'location':
+            // returns a set of scales (xScale, yScale) on every zoom event
+            return self.onLocationChange(viewId, callback, callbackId);
+
+          case 'mouseMoveZoom':
+            return apiPubSub.subscribe('mouseMoveZoom', callback);
+
+          case 'rangeSelection':
+            return apiPubSub.subscribe('rangeSelection', callback);
+
+          case 'viewConfig':
+            return self.onViewChange(callback);
+
+          default:
+            return undefined;
+        }
+      },
+    }
```

Comment:
Yeah the problem is that I had to indent almost every line so it looks like a lot of changes. The core things that are different now are: https://github.com/higlass/higlass/blob/flekschas/pub-sub-provider/app/scripts/api.js#L31-L34

Previously the API created a global pubSub for it's own listeners. In order to avoid this I've inlined the pubSub creation (https://github.com/higlass/higlass/blob/flekschas/pub-sub-provider/app/scripts/api.js#L20-L27). This required me to export "2 API" a private one that is only used internally by HGC and a public one that is being exported to be consumed by some user. Here's the related difference in HGC: https://github.com/higlass/higlass/blob/flekschas/pub-sub-provider/app/scripts/api.js#L20-L27

Does this make sense?

Reply:
I'll take your word for it.

Reply:
Is the difference between the private and public API documented somewhere?

Reply:
No. Where and how do you want internal development things to be documented? This is not a user facing aspect.

Reply:
OK, I didn't realize that it wasn't user-facing. I was under the impression that `api.js` was generally meant for public facing functions.